File: randology-dump-master/thesis/src/uk/ac/stir/randology/generator/GameRand.java
package uk.ac.stir.randology.generator;

// GameRand is an incredibly fast random number generator due to Stephan Schaem.
// see http://www.redditmirror.cc/cache/websites/mjolnirstudios.com_7yjlc/mjolnirstudios.com/IanBullard/files/79ffbca75a75720f066d491e9ea935a0-10.html
public class GameRand implements RNG {
	
	private long low;
	private long high;
	
	public GameRand(long seed) {
		high = seed;
		low = seed ^ 0xDEAFBABE49616E42L;
	};
	
	public long nextLong() {
		high = (high << 32) + (high >> 32);
	    high += low;
	    low += high;
	    return high;
	}
	
    public long next64bits() {
    	return nextLong();
    }
	
	public double nextDouble() {
	    return Math.abs( (double)nextLong() / Long.MAX_VALUE );
	}
	
	@Override
	public String getName() {
		return "GameRand";
	}

}

File: python-implementation/src/generator/GameRand.py
from src.generator.RNG import RNG


class GameRand(RNG):
    def __init__(self, seed: int):
        super().__init__("GameRand")
        self.high = seed
        self.low = seed ^ 0xDEAFBABE49616E42

    def next_long(self):
        self.high = (self.high << 32) + (self.high >> 32)
        self.high = (self.high + self.low) % (2**63 - 1)
        self.low = (self.low + self.high) % (2**63 - 1)
        return self.high

    def next_double(self):
        return abs(self.next_long() / (2 ** 63 - 1))

    def next_64_bits(self):
        return self.next_long()
